File: J9Ch/src/J_8_Data_Structure/Related_7/python_cpp/_541_Zigzag_Iterator_II.cpp

class ZigzagIterator2 {
private:
    int k, p;
    vector<vector<int>::iterator> begins, ends;

public:
    /**
     * @param vecs a list of 1d vectors
     */
    ZigzagIterator2(vector<vector<int>>& vecs) {
        // initialize your data structure here.
        int n = vecs.size();
        k = 0, p = 0;
        for (int i = 0; i < n; ++i) {
            if (vecs[i].size() > 0) {
                k ++;
                begins.push_back(vecs[i].begin());
                ends.push_back(vecs[i].end());
            }
        }
    }

    int next() {
        // Write your code here
        int elem =  *begins[p] ++;
        if (begins[p] == ends[p]) {
            begins.erase(begins.begin() + p);
            ends.erase(ends.begin() + p);
            k --;
            if (k > 0)
                p %= k;
        } else
            p = (p + 1) % k;
        return elem;
    }

    bool hasNext() {
        // Write your code here
        return k > 0;
    }
};

/**
 * Your ZigzagIterator2 object will be instantiated and called as such:
 * ZigzagIterator2 solution(vecs);
 * while (solution.hasNext()) result.push_back(solution.next());
 * Ouptut result
 */

File: J9Ch/src/J_8_DS/other/Iterator/_541_Zigzag_Iterator_II.java
package J_8_DS.other.Iterator;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**541 Zigzag Iterator II

 * Created by tianhuizhu on 7/10/17.
 */

/**
 * Your ZigzagIterator2 object will be instantiated and called as such:
 * ZigzagIterator2 solution = new ZigzagIterator2(vecs);
 * while (solution.hasNext()) result.add(solution.next());
 * Output result
 */


public class _541_Zigzag_Iterator_II {

    public class ZigzagIterator2 {

        public List<Iterator<Integer>> its;
        public int turns;

        /**
         * @param vecs a list of 1d vectors
         */
        public ZigzagIterator2(ArrayList<ArrayList<Integer>> vecs) {
            // initialize your data structure here.
            this.its = new ArrayList<Iterator<Integer>>();
            for (List<Integer> vec : vecs) {
                if (vec.size() > 0)
                    its.add(vec.iterator());
            }
            turns = 0;
        }

        public int next() {
            // Write your code here
            int elem = its.get(turns).next();
            if (its.get(turns).hasNext())
                turns = (turns + 1) % its.size();
            else {
                its.remove(turns);
                if (its.size() > 0)
                    turns %= its.size();
            }
            return elem;
        }

        public boolean hasNext() {
            // Write your code here
            return its.size() > 0;
        }
    }



//-------------------------------------------------------------------------//////////////////





//-------------------------------------------------------------------------//////////////////





//-------------------------------------------------------------------------//////////////////





//-------------------------------------------------------------------------//////////////////





//-------------------------------------------------------------------------//////////////////







}
